src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -3098,6 +3332,802 @@ public static Vector128<float> Invoke(Vector128<float> x, Vector128<float> y)
+            static abstract void Invoke(ref Vector256<float> result, Vector256<float> current, ref Vector256<int> resultIndex, Vector256<int> curIndex, Vector256<float> remainderMask);
+#if NET8_0_OR_GREATER
+            static abstract int Invoke(Vector512<float> result, Vector512<int> resultIndex);
+            static abstract void Invoke(ref Vector512<float> result, Vector512<float> current, ref Vector512<int> resultIndex, Vector512<int> curIndex);
+            static abstract void Invoke(ref Vector512<float> result, Vector512<float> current, ref Vector512<int> resultIndex, Vector512<int> curIndex, Vector512<float> remainderMask);
+#endif
+
+            static virtual float IdentityValue => throw new NotSupportedException();
+        }
+
+        /// <summary>MathF.Max(x, y) (but NaNs may not be propagated)</summary>
+        private readonly struct IndexOfMaxOperator : IIndexOfOperator
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static int Invoke(ReadOnlySpan<float> result)
+            {
+                float curMax = result[0];
+                int curIn = 0;
+                if (float.IsNaN(curMax))
+                {
+                    return curIn;
+                }
+
+                for (int i = 1; i < result.Length; i++)
+                {
+                    float current = result[i];
+                    if (float.IsNaN(current))
+                    {
+                        return i;
+                    }
+
+                    if (curMax == current)
+                    {
+                        if (IsNegative(curMax) && !IsNegative(current))
+                        {
+                            curMax = current;
+                            curIn = i;
+                        }
+                    }
+                    else if (current > curMax)
+                    {
+                        curMax = current;
+                        curIn = i;
+                    }
+                }
+
+                return curIn;
+            }
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public static int Invoke(Vector128<float> result, Vector128<int> maxIndex)
+            {
+                float curMax = float.MinValue;
+                int curIn = 0;
+                for (int i = 0; i < Vector128<float>.Count; i++)
+                {
+                    if (result[i] == curMax && IsNegative(curMax) && !IsNegative(result[i]))
+                    {
+                        curMax = result[i];
+                        curIn = maxIndex[i];
+                    }
+                    else if (result[i] > curMax)
+                    {
+                        curMax = result[i];
+                        curIn = maxIndex[i];
+                    }
+                }
+                return curIn;
```

Comment:
This will do the operation four times.  Could we instead compare result[0]/result[1] and result[2]/result[3], and then compare the result of those two operations to make it 3 instead of 4?

Same should be doable for Vector256, doing four-pairwise comparisons, then two more, and then a final one, giving you 7.  But you should also be able to achieve that just by delegating with GetLower and GetUpper to the Vector128 implementation and then just comparing the results of those.  Similarly for the Vector512.

Seems like you should also be able to factor out the body of this loop into an aggressively inlined helper that you use from all of the vector widths.

That's all unless Tanner has a good idea for how to vectorize this horizontal aggregation.

Reply:
I was trying to figure out how to do the comparison as you are suggesting, but since I need the value and the index I decided to leave it like this for now until I finished the NetStandard code. Then i'll come back and revisit it.

Since it only happens at the very end/1 time do you think it will end up making a big perf difference?

Reply:
For `Arm64` you want `AdvSimd.MaxAcross`.

For `x64` the "best" approach is to shuffle and compare `log2(count)` times. That is, you start with `x = [10, 20, 30, 40]`:
```
y = Shuffle(x, Vector128.Create(2, 3, 0, 1))   // y = [30, 40, 10, 20]
x = Max(x, y)                                  // x = [30, 40, 30, 40]
y = Shuffle(x, Vector128.Create(1, 0, 3, 2))   // y = [40, 30, 40, 30]
x = Max(x, y))                                 // x = [40, 40, 40, 40]
return x.ToScalar();
```

Reply:
> For x64 the "best" approach is to shuffle and compare log2(count) times

@tannergooding, could you look at the other aggregations being performed for other operators, in particular the HorizontalAggregate helper:
https://github.com/dotnet/runtime/blob/bf10b617cbc6b7175b00b1a68599f2597c949710/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs#L2866-L2884
and help fix them if there's meaningful opportunity to do better? Thanks.

Reply:
@tannergooding how did you know which numbers to use in the `Vector128.Create`? And how would you then be able to take that and apply it to the index?

Reply:
The signature for shuffle is `public static Vector128<float> Shuffle (Vector128<float> vector, Vector128<int> indices);`, so the values passed to `Vector128.Create(...)` are simply the indice of `vector` that you want in the particular position.

For `Vector128` we simply want to ensure that every value and so the indices picked don't matter too much, as long as they eventually add up. I tend to start with simply swapping the upper/lower halves, and then swapping pairs for the second. Thus if the default indice order is `[0, 1, 2, 3]`, I choose `[2, 3, 0, 1]`, and then `[1, 0, 3, 2]`.

For `Vector256` and `Vector512` you basically want to do the upper/lower halves and then do the 128-bit algorithm. So for `MaxAcross` for `Vector512` is:
```
// Max the upper/lower halves of the Vector512
Vector256<float> val256 = Max(val512.GetLower(), val512.GetUpper());

// Max the upper/lower halves of the Vector256
Vector128<float> val128 = Max(val256.GetLower(), val256.GetUpper());

// The standard 128-bit algorithm given above
Vector128<float> tmp128 = Shuffle(val128, Vector128.Create(2, 3, 0, 1))
val128 = Max(val128, tmp128);
tmp128 = Shuffle(val128, Vector128.Create(1, 0, 3, 2));
val128 = Max(val128, tmp128);
return val128.ToScalar();
```